### ISSUE TITLE
Move webfonts handling to internal only (not accessible) - WP 6.0 Stopgap

### ISSUE THREAD

**Issue by hellofromtonya** [WordPress/gutenberg#40472]
## What problem does this address?

Part of WordPress 6.0 is style variations to empower themers to define webfonts in their `theme.json`.  Twenty Twenty-Two (TT2) introduces different styling options within its `theme.json`. Within this scope, `theme.json` defines its webfonts  and needs mechanisms to (a) expose the fonts to the typography pickers for user customizations and (b) generate the `@font-face` declarations.

However, the current Webfonts API is still in development and not ready for inclusion in WordPress 6.0. It needs more discussion, development, testing, and feedback to mature before it's ready. There are concerns of rushing the API into Core before its ready as it will lock the API causing future needs limitations, backwards-compatibility issues, high maintenance burden, performance concerns (enqueuing only what is needed, handling of duplicate font variants, etc.), and plugin limitations for further customization (such as changing the font's source).

Style variations don't necessarily need a mature public API at this point. Rather, it needs the mechanisms to (a) wire fonts defined in `theme.json` to the pickers and (b) generate the font-face styles. 

## What is your proposed solution?

Provide a stopgap solution for 6.0 where:

* only the theme's `theme.json` file defines webfonts.
* move the webfonts handling mechanisms to be internal, private functionality within a theme.json handler.
* prevent interaction with webfonts handling mechanisms in such a way that this code is _not accessible_ outside of its container (such as closure or class).
* continue working on Webfonts API  within Gutenberg until it matures and is ready for Core in a future release.

### Keeping the internals really private through abstraction or within a private container

From @peterwilsoncc:
>I’d be pretty happy with an entirely private API for themes.json — not `@private` but actually private in a closure, a private final class, trust us this will be hard deprecated kind of way:
```php
function(){
  $fn = function(){};
}
```
>Basically go to some lengths to ensure that contributors don’t actually end up having to maintain two apis: the temporary one and the permanent one.  Extenders often ignore `@private` (case in point, the million odd sites using a [cron private function](https://wpdirectory.net/search/01G12GPS6NT39F0VQ1M71YTBE6)).

### Empowering ongoing API development in Gutenberg

A mechanism (such as a constant or filter) will be needed to disable this stopgap / temporary implementation once it lands in Core.  Then Gutenberg can turn off Core's internal webfonts handling to use the Webfonts API.

## Props
Props to @desrosj and @peterwilsoncc for [proposing a pared down stopgap approach](https://wordpress.slack.com/archives/C035NHLFUSY/p1650402803393189).

## Tasks

- [x] Identify the essential pieces from the API to support this stopgap. 
- [x] Provide the means to switch between the stopgap version and the API. 
- [x] Move all webfonts handling into a private / unaccessible container.
- [x] Remove all code not related to the `theme.json` file.

## Status

- [X] Add an internal-only theme.json webfonts handler that prevents BC breaks when the API is backported to Core (future). PR #40493 (but do not merge this into Gutenberg - backport directly to WP Core) 🟢  Done ✅ 
- [x] Backport the internal-only theme.json webfonts handler to Core https://github.com/WordPress/wordpress-develop/pull/2622 🟢  Done ✅  with commit https://core.trac.wordpress.org/changeset/53282
- [x] Allow Gutenberg's Webfonts API to disable the stopgate internal-only handler https://github.com/WordPress/gutenberg/pull/40555 🟢  Done ✅

**Comment by hellofromtonya** [comment]
Current status: PR #40493 is ready for review and testing.

**Comment by hellofromtonya** [comment]
Status Update:

1. [As agreed upon by @gzilio, @adamziel, and me](https://wordpress.slack.com/archives/C037DD0CLHY/p1650617641163939),PR #40493 will not be merged into Gutenberg. Why? Only Core needs the stopgap, not the plugin.
2. The backport PR is created https://github.com/WordPress/wordpress-develop/pull/2622.
3. A PR to turn off the stopgap in Core is coming for Gutenberg, i.e. allowing the Webfonts API to run instead of the stopgap in Core.

**Comment by hellofromtonya** [comment]
Update: The stopgap implementation has been committed into WordPress Core's `trunk` ahead of 6.0 Beta 3 via commit https://core.trac.wordpress.org/changeset/53282.

All tasks are completed. Closing this issue.

Thank you everyone for contributing!

**Comment by sabernhardt** [comment]
Follow-up: [Trac 58454](https://core.trac.wordpress.org/ticket/58454) proposes to include `optional` as a valid/allowed `font-display` value.
